[ILM] More resilient when a policy is added to searchable snapshot

gmarouli opened the pull request:
In this PR we implement the proposal 2 of the type inconsistency issue as described in #102705 .

> We propose the following changes in all relevant steps:
If the index is already a searchable snapshot, we retrieve the snapshot information by the lifecycle state but when not present we fallback to the index metadata. We do not update the lifecycle state, this way at the delete phase we will not delete the snapshot. The idea behind this is that we cannot guarantee that the policy created the snapshot, so the policy won't delete it.

In other words, ensures that ILM will be able to proceed in the following scenarios:

- An ILM policy with the a searchable snapshot action in **hot or cold** is added on a **partially mounted** searchable snapshot.
- An ILM policy with the a searchable snapshot action in **frozen** is added on a **fully mounted** searchable snapshot.

The searchable snapshot could have had a previous ILM policy that has been removed via `POST <index>/_ilm/remove` or it might not have been managed at all.

Currently, ILM gets stuck in the above scenarios because:
1) It checks if the searchable snapshot can be skipped, it does not skip because the type of the existing searchable snapshot and the searchable snapshot required by the policy does not match.
2) It check if it needs to generate a snapshot, it does try to generate the snapshot because there are no snapshot data in the lifecycle execution state.
3) It generates a snapshot of the searchable snapshot.
4) When it tries to mount it, it fails because the snapshot contains the searchable snapshot and not the source index.

Relates to https://github.com/elastic/elasticsearch/issues/102705

Fixes https://github.com/elastic/elasticsearch/issues/101958

gmarouli committed:
Add a test to capture the issue

gmarouli committed:
Refactor

gmarouli committed:
Use the existing snapshot metadata

gmarouli committed:
Test simplification

gmarouli committed:
Update docs/changelog/102741.yaml

elasticsearchmachine commented:
Hi @gmarouli, I've created a changelog YAML for you.

gmarouli committed:
Make the test more reliable

gmarouli committed:
Fix format

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix-ilm-resuming-on-searchable-snapshot

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix-ilm-resuming-on-searchable-snapshot

gmarouli committed:
Fix conditions

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

andreidan commented on x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java:
```diff
@@ -574,6 +572,117 @@ public void testConvertingSearchableSnapshotFromFullToPartial() throws Exception
+            assertThat(stepKeyForIndex.name(), is(PhaseCompleteStep.NAME));
+        }, 30, TimeUnit.SECONDS);
+
+        // Ensure the searchable snapshot is not deleted when the index was deleted because it was not created by this
+        // policy. We add the delete phase now to ensure that the index will not be deleted before we verify the above
+        // assertions
```
Sorry for the random comment here - but I'm a bit unsure of this invariant. I'm wondering if it creates unexpected behaviour. The delete phase already has a flag that says "delete snapshot" which has the purpose of allowing our users to delete or skip the deletion of the searchable snapshot that backs a mounted index. 

This new behaviour ignores that flag (in some cases). I don't think we should, I think we should delete the snapshot if the `delete_searchable_snapshot` flag is `true`) https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-delete.html#ilm-delete-options

gmarouli replied:
Hm, that's not how interpreted the docs:

> Deletes the searchable snapshot **created** in a previous phase

Since we did not create the snapshot while executing this policy, we are not going to delete it. The same happens now when we skip the searchable snapshot action and the snapshot was not created by this policy (aka the lifecycle execution does not have a record of this snapshot).

I chose not to delete it for the following reasons:
- It feels to me more inline with the documentation, assuming the `created` word has the significance I thought it had.
- This is how skip searchable snapshot works now when the snapshot was not created during this policy execution.
- And finally, in the scenario that a user has made a backup of their index and then decided to mount it and add an ILM policy to it, we won't delete the backup itself but just the searchable snapshot index. I think some users might see it as an unwanted side-effect.

I understand we might have unwanted searchable snapshots in this way, but the user can always delete them later, but they cannot retrieve the lost data if we delete them by accident.

Does this make sense to you?

andreidan replied:
++ thanks for the details. Let's keep it as is and update the docs to reflect that some snapshots could remain dangling.

andreidan reviewed (commented):
Thanks for working on this Mary

I've left a few minor comments and a question. Can you also please document (in the PR description) which use case this is fixing (i.e. index is fully mounted, ILM policy is removed via `_ilm/remove`, another ILM policy is added that contains frozen is configured for index)

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -411,4 +426,17 @@ public boolean equals(Object o) {
+        return new SearchableSnapshotMetadata(indexName, repo, snapshotName, partial);
+    }
+
+    record SearchableSnapshotMetadata(String indexName, String repositoryName, String snapshotName, boolean partial) {};
```
I like `sourceIndex`!

gmarouli commented on x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java:
```diff
@@ -574,6 +572,117 @@ public void testConvertingSearchableSnapshotFromFullToPartial() throws Exception
+            ),
+            new Phase(
+                "delete",
+                TimeValue.ZERO, // give time for the checks to happen
```
Yes, sharp!

gmarouli commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -234,8 +237,20 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey, XPac
+                String snapshotIndexName;
+                String snapshotName;
+                String repoName;
+                if (lifecycleExecutionState.snapshotName() != null) {
+                    snapshotIndexName = lifecycleExecutionState.snapshotIndexName();
+                    snapshotName = lifecycleExecutionState.snapshotName();
+                    repoName = lifecycleExecutionState.snapshotRepository();
+                } else {
+                    snapshotIndexName = searchableSnapshotMetadata.indexName;
+                    snapshotName = searchableSnapshotMetadata.snapshotName;
+                    repoName = searchableSnapshotMetadata.repositoryName;
+                }
```
Good idea. I will add it.

gmarouli committed:
Improvements based on review

gmarouli committed:
Fix format

gmarouli committed:
Add warning about dangling snapshots

gmarouli commented:
@andreidan thank you for the review. I addressed the comments so far, let me know what you think. I will add a link to the docs preview as soon as it becomes available.

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix-ilm-resuming-on-searchable-snapshot

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java:
```diff
@@ -234,22 +238,36 @@ public List<Step> toSteps(Client client, String phase, StepKey nextStepKey, XPac
-                    lifecycleExecutionState.snapshotRepository(),
-                    lifecycleExecutionState.snapshotIndexName(),
-                    index.getName()
+                logger.info(
```
Should we keep this at debug? (all steps are retryable so wouldn't want to fill the logs with this)

andreidan commented on x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/actions/SearchableSnapshotActionIT.java:
```diff
@@ -574,6 +572,113 @@ public void testConvertingSearchableSnapshotFromFullToPartial() throws Exception
     }
 
+    @SuppressWarnings("unchecked")
+    public void testResumingSearchableSnapshotFromFullToPartial() throws Exception {
```
Should we also add a test for the other use case this fixes - `An ILM policy with the a searchable snapshot action in hot or cold is added on a partially mounted searchable snapshot.` ?

gmarouli replied:
Good idea, I thought it is more or less captured but it's better to be thorough

andreidan reviewed (approved):
LGTM, thanks Mary

gmarouli committed:
Add test case to capture from partially to fully

gmarouli committed:
Fix format

gmarouli committed:
Improve docs

Co-authored-by: Andrei Dan <andrei.dan@elastic.co>

gmarouli committed:
Switch to debug
